clippy_lints/src/loops.rs

```diff
@@ -541,13 +584,266 @@ fn check_for_loop<'a, 'tcx>(
     pat: &'tcx Pat,
     arg: &'tcx Expr,
     body: &'tcx Expr,
-    expr: &'tcx Expr
+    expr: &'tcx Expr,
 ) {
     check_for_loop_range(cx, pat, arg, body, expr);
     check_for_loop_reverse_range(cx, arg, expr);
     check_for_loop_arg(cx, pat, arg, expr);
     check_for_loop_explicit_counter(cx, arg, body, expr);
     check_for_loop_over_map_kv(cx, pat, arg, body, expr);
+    detect_manual_memcpy(cx, pat, arg, body, expr);
+}
+
+fn same_var<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &Expr, var: DefId) -> bool {
+    if_let_chain! {[
+        let ExprPath(ref qpath) = expr.node,
+        let QPath::Resolved(None, ref path) = *qpath,
+        path.segments.len() == 1,
+        // our variable!
+        cx.tables.qpath_def(qpath, expr.id).def_id() == var
+    ], {
+        return true;
+    }}
+
+    false
+}
+
+struct Offset {
+    value: String,
+    negate: bool,
+}
+
+impl Offset {
+    fn negative(s: String) -> Offset {
+        Offset {
+            value: s,
+            negate: true,
+        }
+    }
+
+    fn positive(s: String) -> Offset {
+        Offset {
+            value: s,
+            negate: false,
+        }
+    }
+}
+
+struct FixedOffsetVar {
+    var_name: String,
+    offset: Offset,
+}
+
+fn is_slice_like<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, ty: Ty) -> bool {
+    let is_slice = match ty.sty {
+        ty::TyRef(_, ref subty) => is_slice_like(cx, subty.ty),
+        ty::TySlice(..) => true,
+        ty::TyArray(..) => true,
```

Comment:
rewrite as

```rust
        ty::TySlice(..) | ty::TyArray(..) => true,
```

(this was caught by travis. You can check such things yourself by running `cargo test dogfood`)